### PR TITLE
Keep finger badges inside narrow and short falling notes

### LINKED ISSUE #124: [UI] 검은건반 노트의 운지 번호가 노트보다 넓게 그려져 가독성이 떨어진다
## 현재 상태 · 2026-09-06

**활성 낙하 노트의 검은건반 운지 뱃지 기하는 아직 미해결입니다.** PR #140은 지속음 꼬리에서 운지 표시를 제거했지만, 이 이슈가 지적한 활성 노트의 폭·높이·폰트 하한은 별도입니다.

현재 코드에는 뱃지 최소 14px, 글자 최소 12px 규칙이 있고 짧은 노트/좁은 lane에 따른 충분한 제약이 없습니다. 최초 측정의 10.8~24.5px 흰건반 폭 구간에서 검은건반 뱃지가 넘치는 문제를 다룹니다. 아래 측정표는 과거 측정값으로 보존하며 이번 현행화에서 새 시각 검증을 했다는 뜻은 아닙니다.

### 남은 작업과 완료 조건

- [ ] 뱃지가 자기 노트 폭을 넘지 않도록 하거나, 의도적으로 넘는 배치와 이웃 노트 겹침 규칙을 결정·기록한다.
- [ ] 흰건반 폭 10.8/24/24.46/36px에서 흰·검은 뱃지 기하를 순수 함수 회귀로 검증한다.
- [ ] 0.1초의 짧은 노트에서도 세로 범위를 벗어나지 않도록 clamp 또는 표시 기준을 정의·검증한다.
- [ ] 좁은 화면에서 읽을 수 없는 경우의 표시 여부/크기 기준을 명시한다.
- [ ] 데스크톱과 모바일 가로에서 실제 악보의 가독성을 확인하고 관련 Jest/typecheck/lint/build를 통과한다.

세로 공간 배분·악보 패널은 #125, 운지 숫자 자체의 정확도는 별도 이슈 범위입니다.

근거: PR #140, `src/utils/visualUtils.ts`, `src/components/animation/FallingNotes.tsx`.

<details>
<summary>최초 제보·과거 분석 원문 (이전 수치·정책 설명 포함)</summary>

> 아래 내용은 이력 보존용입니다. 현행 상태와 완료 조건은 위 본문을 따릅니다. 기존 댓글은 그대로 보존했습니다.

## 배경

이슈 #103·#120으로 모든 재생 노트에 운지 번호가 표시된다. 그러나 화면에서 검은건반 노트의 번호는 흰건반보다 읽기 어렵다. 원인은 색이나 폰트가 아니라 **뱃지 상자가 노트 폭과 무관한 픽셀 하한을 갖는 것**이다.

관련 코드는 세 곳이다.

- `src/utils/visualUtils.ts:49` — 노트 폭: 흰건반 `keyPos.w * 0.92`, 검은건반 `keyPos.w * 0.75`
- `src/utils/visualUtils.ts:117` — 뱃지 크기: `min(20, max(14, note.w * 0.7))`
- `src/components/animation/FallingNotes.tsx:56` — 글자 크기: `max(12, badgeSize * 0.8)`

검은건반의 건반 폭 자체가 흰건반의 `13.7/23.5 ≈ 0.583`배(D-037)이므로, 검은건반 노트 폭은 흰건반 폭 대비 `0.583 × 0.75 ≈ 0.437`배다.

## 측정

`BLACK_KEY_WIDTH_RATIO`와 위 세 식을 그대로 계산한 값이다(단위 px).

| 흰건반 폭 | 흰건반 노트 / 뱃지 / 글자 | 검은건반 노트 / 뱃지 / 글자 | 검은건반 뱃지 초과 폭 |
|---:|---|---|---:|
| 10.8 (폰 세로) | 9.94 / 14.00 / 12.0 | 4.72 / 14.00 / 12.0 | **+9.28** |
| 24.0 (`BASE_PLAYBACK_KEY_WIDTH`) | 22.08 / 15.46 / 12.4 | 10.49 / 14.00 / 12.0 | **+3.51** |
| 24.46 (D-021 데스크톱 실측) | 22.50 / 15.75 / 12.6 | 10.69 / 14.00 / 12.0 | **+3.31** |
| 36.0 | 33.12 / 20.00 / 16.0 | 15.74 / 14.00 / 12.0 | -1.74 |

즉 실제로 쓰이는 밀도 구간(10.8~24.5px) 전체에서 검은건반 뱃지는 자기 노트보다 넓다. 흰건반은 같은 구간에서 항상 노트 안에 들어간다.

## 현재 문제

1. **검은건반 뱃지가 노트 밖으로 넘친다.** 기준 밀도에서 좌우로 각 1.75px, 폰 세로에서는 각 4.6px씩 노트 밖에 그려진다. 노트 `div`에 `overflow`가 없어 잘리지는 않지만, 검은건반 노트의 `z-index`가 30(흰건반 노트 20)이므로 **이웃 흰건반 노트 위에 겹쳐 그려진다**.
2. **글자 하한 12px이 검은건반에서만 상시 걸린다.** 검은건반 노트는 10.5px 폭에 12px 글자를 담게 되어 숫자 획이 노트 색 경계와 붙는다. 흰건반은 비례 계산 결과가 하한 위에 있어 이 문제가 없다.
3. **뱃지에 배경판이 없다.** 흰색 굵은 글자 + `textShadow`만 있어서, 좁은 검은건반 노트에서는 글자 주위에 확보되는 노트 색 면적이 거의 없다.
4. **짧은 노트에서는 세로로도 넘친다.** 위치가 `y + h - size - 4`이므로 `h < size + 4`인 노트(140px/s 기준 약 0.13초 미만, 빠른 곡의 16분음표)에서는 뱃지가 노트 위쪽 밖으로 나간다.

## 기대 동작

- 검은건반 노트의 운지 번호가 흰건반과 같은 수준으로 읽힌다.
- 뱃지가 자기 노트의 폭·높이를 벗어나 이웃 노트를 가리지 않는다. 벗어나야 한다면 그것이 의도된 배치(예: 검은건반 전용 세로 오프셋, 배경판 부여)여야 하고 겹침 규칙이 정의돼야 한다.
- 폰 세로(흰건반 약 10.8px)에서 번호를 읽을 수 없다면, 억지로 그리는 대신 표시 여부 기준을 명시한다 — 현재 `shouldShowFingerBadge`는 `finger !== undefined`만 보고 크기를 전혀 보지 않는다.

## 작업 범위 제안

- `getFingerBadgePosition`의 크기 규칙을 노트 폭이 아니라 **건반 lane 폭 + 검은/흰 구분**을 입력으로 받게 한다.
- 검은건반 뱃지에 배경판(원형/사각 plate) 또는 세로 오프셋을 주어 좁은 폭에서 대비를 확보한다.
- 짧은 노트에서 뱃지가 노트 밖으로 나가지 않도록 세로 배치에 clamp를 넣거나, 표시 임계값을 정의한다.
- `shouldShowFingerBadge`에 크기 기준을 추가할지 결정하고 그 근거를 남긴다.

## 완료 조건

- ☐ 검은건반 뱃지가 자기 노트 폭을 넘지 않거나, 넘는 배치가 명시적 결정으로 기록되고 겹침이 정의된다.
- ☐ 흰건반 폭 10.8 / 24 / 24.46 / 36px에서 흰·검은 뱃지 기하를 검사하는 순수 함수 회귀 테스트가 있다(레이아웃 없는 jsdom에서 검증 가능하도록 `visualUtils` 수준에서).
- ☐ 짧은 노트(0.1초)에서 뱃지가 노트 세로 범위를 벗어나지 않는다.
- ☐ 데스크톱과 모바일 가로에서 실제 악보로 검은건반 번호 가독성을 눈으로 확인한다.
- ☐ focused Jest, 전체 Jest, typecheck, lint, build가 통과한다.

## 제외 범위

- 세로 공간 배분·건반 높이 변경 (별도 이슈)
- 상단 악보 표시 (별도 이슈)
- 운지 번호 자체의 정확도 (#120)

Related: #103, #120, #58, D-037

</details>

### PR DESCRIPTION
**Pull request by landfill**
## Problem and behavior

Finger labels used a 14px square even when a black note was only 10.49px wide, and short notes placed the label above their own rectangle. Badges now use independently bounded width/height, a dark plate and integer 12–16px monospace digits. Notes below 10px width or 14px height omit only the label; a 0.1s note at 140px/s still displays its number when wide enough.

D-055 explicitly replaces D-038's unconditional short-note display rule. Existing finger data, inference, note geometry, keyboard layout, audio and timing are unchanged. This does not add the score panel from #125 or implement #146.

## Validation

- Regression-first focused tests: 10 failed / 12 passed before implementation; all 22 pass after; review follow-up adds two boundary cases, now 24 focused tests pass.
- Implementation head 47190a0: full Jest 102 suites / 985 tests, lint, TypeScript, production build and both hosted E2E jobs pass. Follow-up e607701 changes tests only; focused 24 tests and TypeScript pass. Latest hosted checks are shown on this PR.
- Actual FallingNotes component rendered with the repository Love Affair fixture and width/duration cases. All 33 rendered badge rectangles and text ranges fit; desktop and 844×390 landscape browser captures inspected.
- Browser verification uses a static component harness, not a full application end-to-end flow. Physical devices and human performance/readability are not claimed.

## Risk and rollback

Very narrow/short notes intentionally stop displaying labels instead of covering neighboring notes. Musical events and finger values remain intact. Revert this PR to restore the previous rendering policy and its overflow.

Closes #124

Evidence: https://github.com/landfill/ClairKeys/blob/main/docs/recovery/validation/2026-09-07-finger-badge.md

Both CodeRabbit findings were addressed: linked main-branch visual records (including landscape DOM bounds) and replaced the sampled keyboard-width expectation with explicit per-lane boundary cases.

### CODE REVIEW

**Comment by landfill** [comment]
@coderabbitai review

File: docs/recovery/phases/ISSUE-124-finger-badge.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+## Objective
+
+Keep active-note finger badges inside their own notes without changing musical data or keyboard geometry.
+
+## Work stages
+
+1. Reproduce overflow at white-key widths 10.8/24/24.46/36px and short durations before implementation.
+2. Apply D-055 rectangular badge geometry, readable font/contrast and display threshold.
+3. Verify actual rendering in desktop and mobile landscape browser viewports using a real-score fixture.
```
[the remark this answers is not in the document]
Comment (landfill) [reply docs/recovery/phases/ISSUE-124-finger-badge.md:14]: 검증 기록은 저장소 규약에 따라 구현 브랜치가 아니라 main에 직접 커밋했습니다. 최초 기록 9107532에 실제 Love Affair fixture의 FallingNotes 렌더, 데스크톱/844×390 캡처와 DOM 측정이 있습니다. 추가 26cc82c에는 모바일 DOM 측정 33/33 통과도 남겼습니다. 단계는 IN_PROGRESS이며 완료로 표시하지 않았습니다.

검증: https://github.com/landfill/ClairKeys/blob/main/docs/recovery/validation/2026-09-07-finger-badge.md
모바일 측정: https://github.com/landfill/ClairKeys/blob/main/docs/recovery/validation/2026-09-07-finger-badge/mobile-dom-metrics.json

정적 실제 컴포넌트 harness임을 명시했고 전체 앱/실기기/휴먼 실측으로 표현하지 않았습니다.

File: src/utils/__tests__/visualUtils.test.ts [hunk 2]
```diff
@@ -232,3 +233,36 @@ describe('visualUtils with fingering', () => {
...
+        { midi: 60, start: 1, duration, finger: 1, hand: 'R' },
+        { midi: 61, start: 1, duration, finger: 2, hand: 'R' },
+      ], 0, 140, 350, layout);
+      for (const note of notes) {
+        const badge = getFingerBadgePosition(note);
+        expect(badge.x).toBeGreaterThanOrEqual(note.x);
+        expect(badge.y).toBeGreaterThanOrEqual(note.y);
+        expect(badge.x + badge.width).toBeLessThanOrEqual(note.x + note.w + 1e-9);
+        expect(badge.y + badge.size).toBeLessThanOrEqual(note.y + note.h + 1e-9);
+        expect(shouldShowFingerBadge(note)).toBe(keyWidth >= 24 && duration >= 0.1);
```
[the remark this answers is not in the document]
Comment (landfill) [reply src/utils/__tests__/visualUtils.test.ts]: e607701에서 표본 keyWidth>=24 추정을 제거했습니다. 구현 조건을 그대로 복제하는 기대식 대신 흰/검은 lane별 명시적 기대값을 사용하고, 11px(흰 노트만 표시)과 23px(둘 다 표시)을 추가해 변환된 폭의 경계 양쪽을 검증합니다. 기존 10.8/24/24.46/36px 및 duration 경계도 유지합니다. Focused 24 tests, typecheck PASS. Production code unchanged.

**Comment by landfill** [comment]
사용자 결정(2026-09-07)에 따라 병합 없이 철회합니다. 노트 내부 배치와 크기 기준 번호 생략은 실제 요구인 기존 운지 번호 유지와 맞지 않습니다. 배지 상자가 노트보다 넓다는 기하적 사실만으로 사용자 가독성 결함을 확정할 근거도 부족합니다. D-055와 생략 구현은 main에 병합하지 않았으며 배포 앱에는 적용되지 않았습니다. 검증 기록은 철회된 접근의 이력으로만 보존합니다. #146은 기존 main 동작을 기준으로 진행합니다.
